tensorflow/core/kernels/segment_reduction_ops.cc

```diff
@@ -256,36 +256,101 @@ class UnsortedSegmentSumOp : public OpKernel {
     Tensor* output = nullptr;
     OP_REQUIRES_OK(context, context->allocate_output(0, output_shape, &output));
     auto output_flat = output->flat_outer_dims<T>();
-    output_flat.setZero();
-
+    if(is_max_){
+        output_flat.setConstant(std::numeric_limits<T>::min());
+    }else{
+        output_flat.setZero();
+    }
     if (data.NumElements() > 0) {
-      auto data_flat = data.shaped<T, 2>({N, data.NumElements() / N});
-      for (int64 i = 0; i < N; ++i) {
-        Index j = internal::SubtleMustCopy(segment_flat(i));
-        OP_REQUIRES(context, FastBoundsCheck(j, output_rows),
-                    errors::InvalidArgument(
-                        "segment_ids", SliceDebugString(segment_ids.shape(), i),
-                        " = ", j, " is out of range [0, ", output_rows, ")"));
-        output_flat.template chip<0>(j) += data_flat.template chip<0>(i);
+        auto data_flat = data.shaped<T, 2>({N, data.NumElements() / N});
+        for (int64 i = 0; i < N; ++i) {
+            Index j = internal::SubtleMustCopy(segment_flat(i));
+            OP_REQUIRES(context, FastBoundsCheck(j, output_rows),
+                        errors::InvalidArgument(
+                                "segment_ids", SliceDebugString(segment_ids.shape(), i),
+                                " = ", j, " is out of range [0, ", output_rows, ")"));
+            reduction(&data_flat, &output_flat, i, j);
```

Comment:
We are adding a function call inside a loop which could slow down the execution. An alternative is to define the two reduction functions separately as inline functions and pass these functions as template arguments to the class.